orttraining/orttraining/test/python/orttraining_transformer_trainer.py

```diff
@@ -0,0 +1,406 @@
+                self.args.max_steps // (len(train_dataloader) // self.args.gradient_accumulation_steps) + 1
+            )
+        else:
+            t_total = int(len(train_dataloader) // self.args.gradient_accumulation_steps * self.args.num_train_epochs)
+            num_train_epochs = self.args.num_train_epochs
+
+        get_lr_this_step = get_linear_schedule_with_warmup(self.args.warmup_steps, t_total, self.args.learning_rate)
+        loss_scaler = LossScaler('loss_scale_input_name', True, up_scale_window=2000)
+
+        def map_optimizer_attributes(name):
+            no_decay_keys = ["bias", "gamma", "beta", "LayerNorm"]
+            no_decay = any(no_decay_key in name for no_decay_key in no_decay_keys)
+            if no_decay:
+                return {"alpha": 0.9, "beta": 0.999, "lambda": 0.0, "epsilon": 1e-6}
+            else:
+                return {"alpha": 0.9, "beta": 0.999, "lambda": 0.01, "epsilon": 1e-6}
+
+        self.model = ORTTrainer(self.model, None,
+            self.model_desc, 
+            "AdamOptimizer",
+            # TODO: how to support AdamW
+            map_optimizer_attributes=None,
+            learning_rate_description=IODescription('Learning_Rate', [1,], torch.float32),
+            device=self.args.device,
+            postprocess_model=postprocess_model,
+            gradient_accumulation_steps=self.args.gradient_accumulation_steps,
+            world_rank=0, world_size=1,     # only support single GPU cases
+            use_mixed_precision=self.args.fp16,
+            allreduce_post_accumulation=True,
+            get_lr_this_step=get_lr_this_step,
+            loss_scaler=loss_scaler,
+            enable_grad_norm_clip=False,
+            _opset_version=12)
+
+        if self.tb_writer is not None:
+            self.tb_writer.add_text("args", self.args.to_json_string())
+
+        # Train!
+        logger.info("***** Running training *****")
+        logger.info("  Num examples = %d", len(train_dataloader.dataset))
+        logger.info("  Num Epochs = %d", num_train_epochs)
+        logger.info("  Instantaneous batch size per GPU = %d", self.args.per_gpu_train_batch_size)
+        logger.info(
+            "  Total train batch size (w. parallel, distributed & accumulation) = %d",
+            self.args.train_batch_size
+            * self.args.gradient_accumulation_steps
+            * (torch.distributed.get_world_size() if self.args.local_rank != -1 else 1),
+        )
+        logger.info("  Gradient Accumulation steps = %d", self.args.gradient_accumulation_steps)
+        logger.info("  Total optimization steps = %d", t_total)
+
+        global_step = 0
+        epochs_trained = 0
+        steps_trained_in_current_epoch = 0
+
+        tr_loss = 0.0
+        logging_loss = 0.0
+        train_iterator = trange(
+            epochs_trained, int(num_train_epochs), desc="Epoch", disable=self.args.local_rank not in [-1, 0],
+        )
+        for epoch in train_iterator:
```

Comment:
Shouldnt the epoch be in the outer loop while the data iterator in the inner for loop?

Reply:
this:
    for step, inputs in enumerate(epoch_iterator): 
is the data iterator in the inner loop of the epoch outer loop